### PR TITLE
SOLR-16178: ZkController#fireEventListeners thread should be shutdown on close

### PR DESCRIPTION
**Pull request by risdenk**
https://issues.apache.org/jira/browse/SOLR-16178

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java [hunk 1]
```diff
@@ -705,6 +709,8 @@ public void preClose() {
     } finally {
       ExecutorUtil.shutdownAndAwaitTermination(customThreadPool);
     }
+
+    ExecutorUtil.shutdownAndAwaitTermination(fireEventListenersThreadPool);
```
Comment (magibney) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java:713]: Do we perhaps want `shutdownNowAndAwaitTermination()` here? At least some related test failures involve submitted tasks in free-floating Thread blocking searcher init; I think `shutdownAndAwaitTermination()` would track those threads, so not _leak_ them, but might nonetheless deadlock in the same way?

Would it be too cute to shutdown the `fireEventListenersThreadPool` via a task submitted to `customThreadPool`, immediately above? I don't think order should matter with this, so "the sooner the better" might be the way to go (unless there's a reason to delay `fireEventListenersThreadPool` shutdown until after the other close tasks are complete). FWIW I think this use would be in keeping with the spirit of the `customThreadPool`.

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java [hunk 2]
```diff
@@ -2754,19 +2760,18 @@ private boolean fireEventListeners(String zkDir) {
...
+            () -> {
+              log.debug("Running listeners for {}", zkDir);
+              for (final Runnable listener : listenersCopy) {
+                try {
+                  listener.run();
+                } catch (Exception e) {
+                  log.warn("listener throws error", e);
+                }
+              }
+              return null;
```
Comment (magibney) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java:2773]: Could we just remove `return null` and let this be a Runnable as opposed to a Callable? There's no return, and no checked exceptions.

**Review by magibney — CHANGES_REQUESTED** [review]
This is really helpful; thanks for diagnosing/fixing! Aside from the couple of questions this looks great I think (marking "request changes" I guess though, because of the outstanding questions).

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java (hunk 1 above)
Comment (risdenk) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java]: Yea I think thats fair. I did this in dd022238b06df8614a4bd52b2e6fee199b69c9a0

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java (hunk 2 above)
Comment (risdenk) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java]: Yup - addressed in dd022238b06df8614a4bd52b2e6fee199b69c9a0

**Review by magibney — APPROVED** [review]

**Comment by madrob** [comment]
Dupe of #805?

**Comment by risdenk** [comment]
Ugh yup thanks @madrob

**Pull request closed without merging**